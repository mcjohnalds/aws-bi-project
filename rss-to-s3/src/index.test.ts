import { handler } from "./index";
import AWS from "aws-sdk";
import ServerMock from "mock-http-server";
import { promises as fs } from "fs";
import { promisify } from "util";
import { execSync } from "child_process";

const terraformOutput = JSON.parse(
  execSync("terraform output -json", { cwd: ".." }).toString("utf-8")
);

const event = {
  url: "http://localhost:9000/rss.xml",
  bucket: terraformOutput.data_lake_bucket_name.value,
  keyPrefix: "rss-to-s3-test/",
};

const server = new ServerMock({ host: "localhost", port: 9000 });

beforeAll(async () => {
  server.on({
    method: "GET",
    path: "/rss.xml",
    reply: {
      status: 200,
      body: await fs.readFile("./rss-sample.xml"),
    },
  });
  await promisify(server.start)();
});

afterAll(() => promisify(server.stop)());

describe("rss-to-s3", () => {
  it("should write RSS data to S3", async () => {
    const result = await handler(event);

    const s3 = new AWS.S3();
    const object = await s3
      .getObject({
        Bucket: terraformOutput.data_lake_bucket_name.value,
        Key: result.key,
      })
      .promise();

    if (!object.Body) {
      throw new Error("object body was empty");
    }

    const data = object.Body.toString("utf-8");
    expect(data).toMatchSnapshot();
  });

  it("should add a prefix the object key", async () => {
    const result = await handler(event);
    expect(result.key).toMatch(/^rss-to-s3-test\//);
  });

  it("should add a csv suffix the object key", async () => {
    const result = await handler(event);
    expect(result.key).toMatch(/.csv$/);
  });

  it("should write to a different object on every invocation", async () => {
    const result1 = await handler(event);
    const result2 = await handler(event);
    expect(result1.key).not.toEqual(result2.key);
  });
});
