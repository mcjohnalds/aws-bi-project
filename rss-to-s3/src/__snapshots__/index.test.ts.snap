// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rss-to-s3 should write RSS data to S3 1`] = `
"\\"title\\",\\"link\\",\\"pubDate\\",\\"content\\",\\"contentSnippet\\",\\"guid\\",\\"isoDate\\"
\\"Covid: Nadhim Zahawi appointed as vaccine rollout minister\\",\\"https://www.bbc.co.uk/news/uk-55115037\\",\\"Sat, 28 Nov 2020 18:14:14 GMT\\",\\"Nadhim Zahawi will oversee distributing the vaccine in England until at least summer 2021.\\",\\"Nadhim Zahawi will oversee distributing the vaccine in England until at least summer 2021.\\",\\"https://www.bbc.co.uk/news/uk-55115037\\",\\"2020-11-28T18:14:14.000Z\\"
\\"Covid: More than 150 arrests at London anti-lockdown protest\\",\\"https://www.bbc.co.uk/news/uk-england-london-55116470\\",\\"Sat, 28 Nov 2020 22:39:59 GMT\\",\\"Police said they expected arrests to continue to rise following the protest in central London.\\",\\"Police said they expected arrests to continue to rise following the protest in central London.\\",\\"https://www.bbc.co.uk/news/uk-england-london-55116470\\",\\"2020-11-28T22:39:59.000Z\\""
`;
